qiskit_nature/second_q/operators/spin_op.py

```diff
@@ -12,538 +12,606 @@
-        np.add.at(coeff_list, indices, self._coeffs)
-        is_zero = np.isclose(coeff_list, 0, atol=atol)
-        if np.all(is_zero):
-            return SpinOp(
-                (
-                    np.zeros((3, 1, self.register_length), dtype=np.int8),
-                    np.array([0], dtype=np.complex128),
-                ),
-                spin=self.spin,
-            )
-        non_zero = np.logical_not(is_zero)
-        new_array = (
-            flatten_array[non_zero]
-            .reshape((np.count_nonzero(non_zero), 3, self.register_length))
-            .transpose(1, 0, 2)
+        Returns:
+            The index ordered operator.
+        """
+        data = defaultdict(complex)  # type: dict[str, _TCoeff]
+        for terms, coeff in self.terms():
+            label, coeff = self._index_order(terms, coeff)
+            data[label] += coeff
+
+        # after successful index ordering, we remove all zero coefficients
+        return self._new_instance(
+            {
+                label: coeff
+                for label, coeff in data.items()
+                if not np.isclose(_to_number(coeff), 0.0, atol=self.atol)
+            }
         )
-        new_coeff = coeff_list[non_zero]
-        return SpinOp((new_array, new_coeff), spin=self.spin)
 
-    def to_list(self) -> List[Tuple[str, complex]]:
-        """Getter for the list which represents `self`
+    def _index_order(self, terms: list[tuple[str, int]], coeff: _TCoeff) -> tuple[str, _TCoeff]:
+        if not terms:
+            return "", coeff
+
+        # perform insertion sorting
+        for i in range(1, len(terms)):
+            for j in range(i, 0, -1):
+                right = terms[j]
+                left = terms[j - 1]
+
+                if left[1] > right[1]:
+                    terms[j - 1] = right
+                    terms[j] = left
+                    coeff *= -1.0
+
+        new_label = " ".join(f"{term[0]}_{term[1]}" for term in terms)
+        return new_label, coeff
 
-        Returns:
-            The list [(label, coeff)]
-        """
-        coeff_list = self._coeffs.tolist()
-        return [(self._generate_label(i), coeff_list[i]) for i in range(len(self))]
-
-    def _generate_label(self, i):
-        """Generates the string description of `self`."""
-        labels_list = []
-        for pos, (n_x, n_y, n_z) in enumerate(self._spin_array[:, i].T):
-            if n_x >= 1:
-                labels_list.append(f"X_{pos}" + (f"^{n_x}" if n_x > 1 else ""))
-            if n_y >= 1:
-                labels_list.append(f"Y_{pos}" + (f"^{n_y}" if n_y > 1 else ""))
-            if n_z >= 1:
-                labels_list.append(f"Z_{pos}" + (f"^{n_z}" if n_z > 1 else ""))
-        if not labels_list:
-            return f"I_{self.register_length - 1}"
-        return " ".join(labels_list)
-
-    @lru_cache(maxsize=128)
     def to_matrix(self) -> np.ndarray:
+        # TODO: use scipy.sparse.csr_matrix() and add parameter `sparse: bool`.
```

Comment:
This method used to cache its result and this has been removed. I checked out of band with @mrossinek as I was not sure whether it was intentional or not. It seems the removal was not intended.

Reply:
The problem is indeed that `dict` is not hashable and since we use that internally, we cannot hash our operator classes. One solutions would be the following:
```python
class SparseLabelOp:
    ...

    def __hash__(self) -> int:
        return hash(repr(self))
```
This would make all operators hashable (again) and enables the caching again